kona/src/main.rs

```diff
@@ -0,0 +1,283 @@
+use std::{collections::HashMap, sync::Arc};
+
+use clap::Parser;
+use eyre::{bail, Result};
+use kona_derive::{
+    online::{AlloyL2ChainProvider, EthereumDataSource},
+    traits::{ChainProvider, L2ChainProvider, OriginProvider, Pipeline, StepResult},
+    types::{BlockInfo, L2BlockInfo, RollupConfig, StageError, OP_MAINNET_CONFIG},
+};
+use reth::{cli::Cli, transaction_pool::TransactionPool};
+use reth_exex::{ExExContext, ExExEvent, ExExNotification};
+use reth_node_api::FullNodeComponents;
+use reth_node_ethereum::EthereumNode;
+use tracing::{debug, error, info, warn};
+
+mod blobs;
+use blobs::ExExBlobProvider;
+
+mod cli_ext;
+use cli_ext::{KonaArgsExt, ValidationMode};
+
+mod pipeline;
+use pipeline::{new_local_pipeline, LocalAttributesBuilder, LocalPipeline};
+
+mod providers;
+use providers::LocalChainProvider;
+
+mod validation;
+use validation::{AttributesValidator, EngineApiValidator, TrustedValidator};
+
+#[derive(Debug)]
+pub(crate) struct KonaExEx<Node: FullNodeComponents> {
+    /// The rollup configuration
+    cfg: Arc<RollupConfig>,
+    /// The context of the Execution Extension
+    ctx: ExExContext<Node>,
+    /// The chain provider to follow the chain state based on ExEx events
+    chain_provider: LocalChainProvider,
+    /// The L2 chain provider to fetch L2 block information and optionally
+    /// verify newly derived payloads against
+    l2_provider: AlloyL2ChainProvider,
+    /// The blob provider to fetch blobs from the beacon client
+    blob_provider: ExExBlobProvider,
+    /// The validator to verify newly derived payloads
+    validator: Box<dyn AttributesValidator + Send>,
+    /// The derivation pipeline
+    pipeline: LocalPipeline,
+    /// The current L2 block we are processing
+    l2_block_cursor: L2BlockInfo,
+    /// A map of L1 anchor blocks to their L2 cursor
+    l1_to_l2_block_cursor: HashMap<BlockInfo, L2BlockInfo>,
+    /// Whether the ExEx is synced to the L2 genesis
+    is_synced_to_l2_genesis: bool,
+    /// Whether we should advance the L2 block cursor
+    should_advance_l2_block_cursor: bool,
+    /// The number of derived payloads so far
+    derived_payloads_count: u64,
+}
+
+impl<Node: FullNodeComponents> KonaExEx<Node> {
+    /// Creates a new instance of the Kona Execution Extension.
+    pub async fn new(ctx: ExExContext<Node>, args: KonaArgsExt) -> Self {
+        // TODO: make chain id configurable
+        let cfg = Arc::new(OP_MAINNET_CONFIG);
```

Comment:
Since @clabby and myself have updated the [superchain-registry](https://github.com/ethereum-optimism/superchain-registry/tree/main), we should now support dynamically loading the chain config via chain id. Let's add a (default `10`) l2 chain id param to `KonaArgsExt`, `--l2-chain-id` maybe.

Kona's `trusted-sync` example [does this the outdated way](https://github.com/ethereum-optimism/kona/blob/main/examples/trusted-sync/src/main.rs#L49-L51) via the `superchain-primitives` crate which should be updated to use the `serde`-enabled `superchain-bindings` crate.

Then here, we can load the rollup config using the [superchain-registry rust bindings](https://github.com/ethereum-optimism/superchain-registry/blob/main/bindings/rust-bindings/src/lib.rs#L30).

```suggestion
        let cfg = superchain_bindings.ROLLUP_CONFIGS.get(&args.l2_chain_id).expect("Missing l2 chain id in superchain bindings.");
        let cfg = Arc::new(cfg);
```

Reply:
Awesome, I see that `kona` and the `superchain-registry` have incompatible versions. 
This seems to be blocked by https://github.com/ethereum-optimism/kona/pull/393